src/main/java/org/folio/service/OverduePeriodCalculatorService.java

```diff
@@ -0,0 +1,247 @@
+  private final FeesFinesClient feesFinesClient;
+  private final InventoryStorageClient inventoryStorageClient;
+
+  public OverduePeriodCalculatorService(Vertx vertx, Map<String, String> okapiHeaders) {
+    this.calendarClient = new CalendarClient(vertx, okapiHeaders);
+    this.circulationStorageClient = new CirculationStorageClient(vertx, okapiHeaders);
+    this.feesFinesClient = new FeesFinesClient(vertx, okapiHeaders);
+    this.inventoryStorageClient = new InventoryStorageClient(vertx, okapiHeaders);
+  }
+
+  public Future<Integer> getMinutes(Loan loan, DateTime systemTime) {
+    if (loan.getOverdueFinePolicyId() == null) {
+      log.error("Loan ID {} - overdue fine policy is missing", loan.getId());
+    }
+
+    return succeededFuture(new CalculationContext())
+      .map(ctx -> ctx.withLoan(loan))
+      .compose(ctx -> feesFinesClient.findOverdueFinePolicyById(loan.getOverdueFinePolicyId())
+        .map(ctx::withOverdueFinePolicy))
+      .compose(ctx -> circulationStorageClient.findLoanPolicyById(loan.getLoanPolicyId())
+        .map(ctx::withLoanPolicy))
+      .compose(ctx -> {
+        if (preconditionsAreMet(ctx, systemTime)) {
+          return succeededFuture(ctx)
+            .compose(r -> getOverdueMinutes(ctx, systemTime)
+              .map(om -> adjustOverdueWithGracePeriod(ctx, om)));
+        }
+        else {
+          return succeededFuture(ZERO_MINUTES);
+        }
+      });
+  }
+
+  private boolean preconditionsAreMet(CalculationContext context, DateTime systemTime) {
+    return context.getOverdueFinePolicy().getCountClosed() != null
+      && loanIsOverdue(context.getLoan(), systemTime);
+  }
+
+  private boolean loanIsOverdue(Loan loan, DateTime systemTime) {
+    return ObjectUtils.allNotNull(loan, systemTime)
+      && loan.getDueDate().before(systemTime.toDate());
+  }
+
+  Future<Integer> getOverdueMinutes(CalculationContext calculationContext, DateTime systemTime) {
+    boolean shouldCountClosedPeriods = calculationContext.getOverdueFinePolicy().getCountClosed();
+    Loan loan = calculationContext.getLoan();
+
+    return shouldCountClosedPeriods || getItemLocationPrimaryServicePoint(loan) == null
+      ? minutesOverdueIncludingClosedPeriods(loan, systemTime)
+      : minutesOverdueExcludingClosedPeriods(loan, systemTime);
+  }
+
+  private Future<Integer> minutesOverdueIncludingClosedPeriods(Loan loan, DateTime systemTime) {
+    int overdueMinutes = minutesBetween(new DateTime(loan.getDueDate()), systemTime).getMinutes();
+    return succeededFuture(overdueMinutes);
+  }
+
+  private Future<Integer> minutesOverdueExcludingClosedPeriods(Loan loan, DateTime returnDate) {
+    DateTime dueDate = new DateTime(loan.getDueDate());
+    return getItemLocationPrimaryServicePoint(loan)
+      .compose(primaryServicePointId -> calendarClient
+        .fetchOpeningDaysBetweenDates(primaryServicePointId.toString(), dueDate, returnDate, false)
```

Comment:
check NPE please

Reply:
Changed this piece - service point ID is passed as a method parameter now.